python/lsst/daf/butler/_butler.py

```diff
@@ -85,29 +104,77 @@ class Butler:
     butler : `Butler`, optional.
         If provided, construct a new Butler that uses the same registry and
         datastore as the given one, but with the given collection and run.
-        Incompatible with the ``config`` and ``searchPaths`` arguments.
+        Incompatible with the ``config``, ``searchPaths``, and ``writeable``
+        arguments.
     collection : `str`, optional
-        Collection to use for all input lookups, overriding
-        config["collection"] if provided.
+        Collection to use for all input lookups.  May be `None` to either use
+        the value passed to ``run``, or to defer passing a collection until
+        the methods that require one are called.
     run : `str`, optional
         Name of the run datasets should be output to; also used as a tagged
         collection name these dataset will be associated with.  If the run
-        does not exist, it will be created.  If "collection" is None, this
+        does not exist, it will be created.  If ``collection`` is `None`, this
         collection will be used for input lookups as well; if not, it must have
-        the same value as "run".
+        the same value as ``run``.
     searchPaths : `list` of `str`, optional
         Directory paths to search when calculating the full Butler
         configuration.  Not used if the supplied config is already a
         `ButlerConfig`.
+    writeable : `bool`, optional
+        Explicitly sets whether the butler supports write operations.  If not
+        provided, a read-only butler is created unless ``run`` is passed.
 
     Raises
     ------
     ValueError
         Raised if neither "collection" nor "run" are provided by argument or
         config, or if both are provided and are inconsistent.
     """
+    def __init__(self, config: Union[Config, str, None] = None, *,
+                 butler: Optional[Butler] = None,
+                 collection: Optional[str] = None,
+                 run: Optional[str] = None,
+                 searchPaths: Optional[List[str]] = None,
+                 writeable: Optional[bool] = None):
+        if butler is not None:
+            if config is not None or searchPaths is not None or writeable is not None:
+                raise TypeError("Cannot pass 'config', 'searchPaths', or 'writeable' "
+                                "arguments with 'butler' argument.")
+            self.registry = butler.registry
+            self.datastore = butler.datastore
+            self._storageClasses = butler.storageClasses
+            self._composites = butler.composites
```

Comment:
Yup, this is poor test coverage in `daf_butler`, though code exercising it in obs_base is tested in `ci_hsc` (both of them, I think), and those did fail - Jenkins just hadn't gotten that far yet when I put this into review.

Instead of adding tests to `daf_butler`, though, I may see if I can just remove the `butler` argument and instead change the `obs_base` code to stop using it - this ticket was originally created because I wanted to make a butler before I knew the run/collection in that code, and I added the `butler` argument so I could quickly make new butlers with different runs/collections as a workaround.  I figure less code to test is even better than more tests.  But if that doesn't work, I'll add a test to `daf_butler`.

Reply:
It doesn't seem crazy to seed a new butler from another butler and then change the collection. A test seems trivial (create the butler then create another from it).